keiko/keiko/keiko.py

```diff
@@ -135,32 +164,41 @@ def generate_report(
 
         # run pdflatex
         cmd = [
-            "pdflatex",
+            "latexmk",
+            "-xelatex",
             "-synctex=1",
             "-interaction=nonstopmode",
             preprocessed_tex_path.as_posix(),
         ]
         env = {**os.environ, "TEXMFVAR": directory}
-        for i in (1, 2):
-            output = subprocess.run(cmd, cwd=directory, env=env, capture_output=True, check=False)
-            current_span.add_event(f"Completed pdflatex run {i}")
+
+        def log_output(level, output):
+            if not logger.isEnabledFor(level):
+                return
+            # prefix all lines in output
+            for line in output.decode("utf-8").splitlines():
+                logger.log(level, "latexmk output: %s", line)
```

Comment:
Just something that popped into my mind. Since only Logger objects can create log records, another way to achieve this is to extend a logger object and implement `log`. E.g.:
```python
class MyLogger(Logger):
    def log(self, level, msg, *args, **kwargs):
        for line in output.decode().splitlines():
            super().log(level, "latexmk output: %s", line)
...
logger = MyLogger()
...
```

Reply:
To me it seems a bit more convoluted to create a new logger class and object for one bit of logging, so I'd prefer to keep the current function.

Reply:
I agree. I also don't really like changing the semantics of `log` from logging one single line to a for loop that logs several lines.

Reply:
Yes but regardless of the suggestion, it also doesn't make sense, or at least not common,  to write multiple log records (because of newlines) instead of one log record with the newlines